### PR TITLE
Lots of functional Kotlin examples

### PR DESCRIPTION
**Pull request by ChrisCanCompute**
Commits include:
 - Update to .gitignore to make setup easier.
 - Addition of the files from functional1 of Kotlin Hours
 - Definition of a fully funcitonal Option type in Kotlin

### CODE REVIEW

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+    abstract fun get(): T
+
+    override fun <R> map(fn: (T) -> R): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            Some(fn(get()))
+        }
+    }
+    override fun <R> apply(fn: Applicative<(T) -> R>): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
+            fn.map { it(get()) } as FunctionalOption<R>
```
Comment (ChrisCanCompute) [inline src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: I suspect there's a declaration that doesn't require a cast here, but I couldn't get the generic types quite right.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+        }
+    }
+    override fun <R> lift(fn: (T) -> Monad<R>): FunctionalOption<R> {
+        return if(isEmpty()) None() else {
+            fn(get()) as FunctionalOption<R>
```
Comment (ChrisCanCompute) [inline src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: I couldn't get the types right here either.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt [hunk 3]
```diff
@@ -0,0 +1,66 @@
+package uk.co.llobile.kotlin.hours.types
+
+sealed class FunctionalOption<T>: Functor<T>, Applicative<T>, Monad<T> {
+
+    abstract fun isEmpty(): Boolean
+    abstract fun get(): T
+
+    override fun <R> map(fn: (T) -> R): FunctionalOption<R> {
+        return if (isEmpty()) None() else {
```
Comment (massimocarli) [inline src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: I think _when_ is more elegant that _if/else_

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt [hunk 4]
```diff
@@ -0,0 +1,66 @@
+package uk.co.llobile.kotlin.hours.types
+
+sealed class FunctionalOption<T>: Functor<T>, Applicative<T>, Monad<T> {
```
Comment (massimocarli) [inline src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: If an Applicative<T> is also a Functor<T> we don't need to use both of them here.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt (hunk 1 above)
Comment (massimocarli) [reply src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: Also here use _when_ please.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    override fun isEmpty() = false
+    override fun get() = value
+
+    override fun toString(): String {
+        return String.format("Some(%s)", value)
+    }
+
+}
+
+class None<T>() : FunctionalOption<T>() {
```
Comment (massimocarli) [inline src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: I don't think this should be a class. It should be a not generic **object**. if you set the `FunctionalOption` as covariant you can set the `None` as `FunctionalOption<Nothing>`

File: src/uk/co/llobile/kotlin/hours/types/Monad.kt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+package uk.co.llobile.kotlin.hours.types
+
+/**
+ * Definition of a Monad.
+ */
+interface Monad<T> {
+    fun <R> lift(fn: (T) -> Monad<R>): Monad<R>
```
Comment (massimocarli) [inline src/uk/co/llobile/kotlin/hours/types/Monad.kt]: Historically the function is called _flatMap_. It should also provide a _unity()_

**Review by massimocarli — CHANGES_REQUESTED** [review]

**Review by massimocarli — CHANGES_REQUESTED** [review]
Please put this into a different pull request.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt (hunk 1 above)
Comment (ChrisCanCompute) [reply src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: Done

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt (hunk 4 above)
Comment (ChrisCanCompute) [reply src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: I've made Applicative not extend Functor, as I think that's more correct.

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt (hunk 3 above)
Comment (ChrisCanCompute) [reply src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: Done

File: src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt (hunk 5 above)
Comment (ChrisCanCompute) [reply src/uk/co/llobile/kotlin/hours/types/FunctionalOption.kt]: Done

File: src/uk/co/llobile/kotlin/hours/types/Monad.kt (hunk 6 above)
Comment (ChrisCanCompute) [reply src/uk/co/llobile/kotlin/hours/types/Monad.kt]: Wikipedia suggests flatMap() and join(). I'll probably go with that.

File: src/uk/co/llobile/kotlin/hours/types/Monad.kt [hunk 7]
```diff
@@ -3,6 +3,8 @@ package uk.co.llobile.kotlin.hours.types
 /**
  * Definition of a Monad.
  */
-interface Monad<T> {
-    fun <R> lift(fn: (T) -> Monad<R>): Monad<R>
+interface Monad<out T> {
+    fun <R> flatMap(fn: (T) -> Monad<R>): Monad<R>
+
+    fun <R> join(): Monad<R>
```
Comment (massimocarli) [inline src/uk/co/llobile/kotlin/hours/types/Monad.kt]: Not sure the `join()` is the same as `unit()`

**Pull request closed without merging**